src/runtime-rs/crates/resource/src/network/network_model/route_model.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use anyhow::{anyhow, Context, Result};
+use async_trait::async_trait;
+use tokio::process::Command;
+
+use super::{NetworkModel, NetworkModelType};
+use crate::network::NetworkPair;
+
+#[derive(Debug)]
+pub(crate) struct RouteModel {}
+
+impl RouteModel {
+    pub fn new() -> Result<Self> {
+        Ok(Self {})
+    }
+}
+
+#[async_trait]
+impl NetworkModel for RouteModel {
+    fn model_type(&self) -> NetworkModelType {
+        NetworkModelType::Route
+    }
+
+    async fn add(&self, pair: &NetworkPair) -> Result<()> {
+        let tap_name = &pair.tap.tap_iface.name;
+        let virt_name = &pair.virt_iface.name;
+        let virt_iface_addr = pair.virt_iface.addrs[0].addr.to_string();
+
+        let commands_args = vec![
+            vec![
+                "rule", "add", "pref", "10", "from", "all", "lookup", "local",
+            ],
+            vec!["rule", "del", "pref", "0", "from", "all"],
+            vec!["rule", "add", "pref", "5", "iif", virt_name, "table", "10"],
+            vec![
+                "route", "replace", "default", "dev", tap_name, "table", "10",
+            ],
+            vec![
+                "neigh",
+                "replace",
+                &virt_iface_addr,
+                "lladdr",
+                &pair.virt_iface.hard_addr,
+                "dev",
+                tap_name,
+            ],
+        ];
+
+        for ca in commands_args {
+            let output = Command::new("ip")
```

Comment:
`/sbin/ip` for security.

Relying on external commands like this is a necessary evil as we clearly don't want to (and should not have to) re-implement their functionality), but the worry is that they might change their external API in unexpected ways and break this code.

As such, I'm wondering if we should log the version details?:

```bash
$ tc -V
$ ip -V
$ iptables -V
```

Since they might also depend on kernel modules, I wonder if we should also perform basic checks on the running guest kernel?

Ideally, we'd only log all these details on failure (to minimise extraneous output at startup, but also to avoid slowing down startup.

I wonder if we could add a function to the `Network` or `NetworkModel` trait for this? Something like:

```rust
// Only called on error to determine details of the network
async fn details(&self, ...) -> Result<String>;
```

As long as we can find a single point in the code where `.details()` could be called when other network operations fail, that would solve the problem of potentially slowing down startup.

Reply:
Thanks for your review, use the command is a workaround, we'll use the netlink API to replace the command.

Reply:
@quanweiZhou - When we switch to netlink, we can remove those commands. But for security, reliability, consistency, _etc_ reasons, we need to run all external commands by specifying their full path.